### PR TITLE
Buffs atmospheric analyzers to show mole count, volume and higher degrees of precision

### PR DESCRIPTION
**Pull request by mirrorcult**
:cl:
add: Analyzers can now scan all kinds of atmospheric machinery - unary, binary, ternary, quaternary, you name it. This means stuff like gas pumps, gas mixers, vents and so forth can be analyzed.
tweak: Analyzers now show temperature in kelvin as well as celsius.
tweak: Analyzers now show total mole count, volume, and mole count of all gases.
tweak: Analyzers show everything at slightly higher degrees of precision.
/:cl:

[why]: More accurate and advanced analyzers have been something that I think atmos and toxins mains have been wanting for a while now (including myself). It's always a pain to have to manually add 273.15 to every celsius temperature, or have to use outside tools or VV (which isnt usable in actual games) to calculate mole count for a gas mixture.  Not to mention, the ability to analyze things like volume pumps and gas filters makes it easier to manage larger atmos setups. This is just a QoL change for people in any of those jobs. In fact, most players probably don't even know what the fuck is a mole is so adding mole readout to analyzers is a good way to get people to understand atmospherics better.

Some images of analyzing different things + icons (if icondiffbot isnt obedient)
Icon is a bit coderspritey but I tried my best alright

https://imgur.com/a/jPU0fsC

### CODE REVIEW

**Comment by optimumtact** [comment]
Do we really need yet another analyzer?

**Comment by cyclowns** [comment]
..since when is that an issue?
there's like.. four different analyzers in the game? t-ray, health, gas, slime? they all have their own separate uses. this PR just adds an advanced version of ONE of them, it's not like i'm adding a FunPack TM of 50 new analyzers

**Comment by optimumtact** [comment]
we don't even need four

**Comment by Iatots** [comment]
Anyone who actually uses an analyzer will use these because they are available and one month from now people will start to ask "why is there a normal analyzer in the first place?", and just like old medscanners, the advanced version will become the standard.

:clap:your:clap:quality:clap:of:clap:life:clap:is:clap:power:clap:creep:clap:

**Comment by NuclearPhone** [comment]
haha yeah lets remove the health analyzer

**Comment by cyclowns** [comment]
@optimumtact ...ok? what do you want to do instead? remove every single analyzer and replace it with one MegaAnalyzer that analyzes everything in the game? all four analyzers are extremely useful and it would be pretty detrimental to remove any of them. doctors need to see patients health, engineers need to easily see cabling/pipes under the floor, atmos techs need to look at gas, and xenobiologists need to see their slimes mutation chances.

should we just go ahead and remove every item that has more than four of itself? should we just remove every single gun in the game in favor of just pistols and eguns? i can kind of see why people would favor just replacing the old analyzer with the advanced version but saying that the amount we have right now is TOO MANY is a stretch.

@Iatots i see what you mean by this kind of being powercreep, but that's sort of the point. advanced analyzers are meant to be the standard for anyone that would use them on a regular basis, and normal analyzers are there for people like miners on lavaland/assistants making backalley bombs that don't need the functionality.

**Comment by cyclowns** [comment]
if more people would rather just have the normal analyzer be replaced by the advanced (without the name change and all), then i'm more than happy to do that

**Comment by optimumtact** [comment]
yes sounds fine to me

**Comment by RandomMarine** [comment]
Can they be made to display container volume too?

**Comment by cyclowns** [comment]
that's definitely possible yeah, I'll probably add that along with displaying total moles (which I forgot)

**Comment by cyclowns** [comment]
apologies for all the commits, I am unintelligent

**Review by Jalleo — CHANGES_REQUESTED** [review]
Why not do this for meters instead and make it so analysers can do it but requires time to do so?

File: code/game/objects/items/devices/scanners.dm [hunk 1]
```diff
@@ -356,13 +358,46 @@ GAS ANALYZER
 	var/cooldown = FALSE
 	var/cooldown_time = 250
 	var/accuracy // 0 is the best accuracy.
+	var/advanced = FALSE //advanced analyzer?
```
Comment (ExcessiveUseOfCobblestone) [inline code/game/objects/items/devices/scanners.dm]: remnant?

Comment (cyclowns) [reply code/game/objects/items/devices/scanners.dm]: whoops, will remove

**Comment by cyclowns** [comment]
@Jalleo that comes with the issue that there would be no way to check the contents of canisters, tanks, or non-pipe components instantly, and depending on the time delay that could be a huge deal

File: code/game/objects/items/devices/scanners.dm (hunk 1 above)
Comment (cyclowns) [reply code/game/objects/items/devices/scanners.dm]: fixed

File: code/game/objects/items/devices/scanners.dm [hunk 2]
```diff
@@ -361,8 +363,40 @@ GAS ANALYZER
...
 
-/obj/item/analyzer/attack_self(mob/user)
+/obj/item/analyzer/afterattack(obj/holder, mob/user, proximity_flags)
+	if(holder.Adjacent(user))
+		//there has to be a better way to do this
+		if (istype(holder, /obj/item/tank))
+			var/obj/item/tank/T = holder
+			atmosanalyzer_scan(T.air_contents, user, T)
+
+		else if (istype(holder, /obj/item/onetankbomb))
```
Comment (Fox-McCloud) [inline code/game/objects/items/devices/scanners.dm]: This is so gross. Why not have `analyzer_act`?

Comment (cyclowns) [reply code/game/objects/items/devices/scanners.dm]: i think its ugly too, i just dont know an objectively better way to do it
care to elaborate?

Comment (cyclowns) [reply code/game/objects/items/devices/scanners.dm]: actually I think I know what you mean
i'll try that out

Comment (cyclowns) [reply code/game/objects/items/devices/scanners.dm]: @Fox-McCloud changed, this is what you meant yes?

Comment (Fox-McCloud) [reply code/game/objects/items/devices/scanners.dm]: Yes, it is. Thank you so much. This is much more OOP and far superior. Really appreciate it.

**Comment by ExcessiveUseOfCobblestone** [comment]
The basic conversion (C > K) should be default but I think higher degrees of precision/more information should require higher-tiered stockparts like the laser. 

Bonus points if you transfer that school of thot to the others.

**Comment by cyclowns** [comment]
@ExcessiveUseOfCobblestone i did do basically that and people complained it was powercreep

**Comment by ExcessiveUseOfCobblestone** [comment]
lol adding a constant is a balance mechanic, what if you type it wrong in ur calculator?????

**Comment by nicbn** [comment]
Volume is always constant in a container
The "code" volume is not real volume
Don't show it up

**Comment by cyclowns** [comment]
@nicbn uhh.. you ok? because the volume of the container the gas mixture is in IS constant (unless its modified by adding a pipe to a pipenet or something). unless you have a problem with that?

**Comment by nicbn** [comment]
Oh I didn't know it showed up the pipenet volume
That's one hundred percent not intuitive

**Comment by cyclowns** [comment]
what else would it show?

**Comment by nicbn** [comment]
The volume of the pipe?

**Comment by cyclowns** [comment]
thats not how it works

**Comment by nicbn** [comment]
Yeah and the way it works

>'s one hundred percent not intuitive

**Comment by cyclowns** [comment]
thats not my problem for this pr

**Comment by RandomMarine** [comment]
How the fuck is volume not intuitive?

It's _less_ intuitive to keep that value hidden from players.

**Pull request merged**